File: android/app/src/main/kotlin/com/musicmuni/calibra/sample/BreathMonitorSection.kt
package com.musicmuni.calibra.sample

import android.content.Context
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import com.musicmuni.calibra.PitchDetector
import com.musicmuni.sonix.SonixRecorder
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

private const val PREFS_NAME = "breath_monitor_prefs"
private const val KEY_BEST_SCORE = "best_score"
private const val RMS_THRESHOLD = 0.01f
private const val SILENCE_GRACE_MS = 500L

/**
 * Breath Monitor Section - Duration tracking with VAD and silence inertia.
 *
 * Flow:
 * 1. User starts monitoring
 * 2. Wait for voice to be detected
 * 3. Start timer when singing begins
 * 4. Allow 500ms silence grace period
 * 5. Stop timer when silence exceeds grace period
 * 6. Compare with best score and persist
 */
@Composable
fun BreathMonitorSection() {
    val context = LocalContext.current
    val scope = rememberCoroutineScope()

    // State
    var monitoringState by remember { mutableStateOf(BreathMonitorState.IDLE) }
    var elapsedSeconds by remember { mutableFloatStateOf(0f) }
    var bestScore by remember { mutableFloatStateOf(loadBestScore(context)) }
    var isVoiceDetected by remember { mutableStateOf(false) }
    var recordingLevel by remember { mutableFloatStateOf(0f) }
    var currentPitch by remember { mutableFloatStateOf(0f) }
    var status by remember { mutableStateOf("Hold a note as long as you can!") }

    // Audio resources
    var recorder by remember { mutableStateOf<SonixRecorder?>(null) }
    var pitchHandle by remember { mutableLongStateOf(0L) }

    // Timing
    var startTimeMs by remember { mutableLongStateOf(0L) }
    var lastVoiceTimeMs by remember { mutableLongStateOf(0L) }

    // Start monitoring
    fun startMonitoring() {
        scope.launch {
            // Reset state
            elapsedSeconds = 0f
            isVoiceDetected = false
            monitoringState = BreathMonitorState.WAITING_FOR_VOICE
            status = "Start singing when ready..."

            // Create audio resources
            val recordPath = "${context.cacheDir}/breath_monitor.m4a"
            recorder?.release()
            recorder = SonixRecorder.create(recordPath, "m4a", "voice")

            if (pitchHandle != 0L) PitchDetector.destroy(pitchHandle)
            pitchHandle = PitchDetector.create(1024)

            // Start recording
            recorder?.start()

            // Collect level for meter
            launch {
                recorder?.level?.collect { level ->
                    recordingLevel = level
                }
            }

            // Main detection loop
            launch {
                recorder?.audioBuffers?.collect { buffer ->
                    if (monitoringState == BreathMonitorState.IDLE ||
                        monitoringState == BreathMonitorState.COMPLETE
                    ) {
                        return@collect
                    }

                    // Detect pitch
                    val pitch = PitchDetector.detect(
                        pitchHandle,
                        buffer.data,
                        buffer.sampleCount,
                        16000,
                        0.15f,
                        80f,
                        1200f,
                        0f,
                        0
                    )
                    currentPitch = pitch

                    // Voice is detected if pitch is valid (> 0) and level is above threshold
                    val hasVoice = pitch > 0 && recordingLevel > RMS_THRESHOLD
                    val currentTimeMs = System.currentTimeMillis()

                    when (monitoringState) {
                        BreathMonitorState.WAITING_FOR_VOICE -> {
                            if (hasVoice) {
                                // Voice detected - start timing
                                monitoringState = BreathMonitorState.SINGING
                                startTimeMs = currentTimeMs
                                lastVoiceTimeMs = currentTimeMs
                                isVoiceDetected = true
                                status = "Keep going!"
                            }
                        }
                        BreathMonitorState.SINGING -> {
                            if (hasVoice) {
                                // Voice still present - update last voice time
                                lastVoiceTimeMs = currentTimeMs
                                isVoiceDetected = true
                            } else {
                                // No voice - check grace period
                                isVoiceDetected = false
                                val silenceDuration = currentTimeMs - lastVoiceTimeMs

                                if (silenceDuration > SILENCE_GRACE_MS) {
                                    // Grace period exceeded - stop
                                    monitoringState = BreathMonitorState.COMPLETE
                                    elapsedSeconds = (lastVoiceTimeMs - startTimeMs) / 1000f

                                    // Update best score if needed
                                    if (elapsedSeconds > bestScore) {
                                        bestScore = elapsedSeconds
                                        saveBestScore(context, bestScore)
                                        status = "New record! ${formatTime(elapsedSeconds)}"
                                    } else {
                                        status = "Good try! ${formatTime(elapsedSeconds)}"
                                    }

                                    recorder?.stop()
                                }
                            }
                        }
                        else -> {}
                    }

                    // Update elapsed time during singing
                    if (monitoringState == BreathMonitorState.SINGING) {
                        elapsedSeconds = (currentTimeMs - startTimeMs) / 1000f
                    }
                }
            }
        }
    }

    // Stop monitoring
    fun stopMonitoring() {
        scope.launch {
            recorder?.stop()
            recorder?.release()
            recorder = null
            monitoringState = BreathMonitorState.IDLE
            status = "Hold a note as long as you can!"
        }
    }

    // Reset best score
    fun resetBestScore() {
        bestScore = 0f
        saveBestScore(context, 0f)
    }

    // Cleanup on dispose
    DisposableEffect(Unit) {
        onDispose {
            recorder?.release()
            if (pitchHandle != 0L) PitchDetector.destroy(pitchHandle)
        }
    }

    // UI
    Column(
        modifier = Modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(12.dp)
    ) {
        Text(
            text = "Breath Monitor",
            style = MaterialTheme.typography.titleMedium
        )

        Text(
            text = status,
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant
        )

        // Main timer display
        Card(
            modifier = Modifier.fillMaxWidth(),
            colors = CardDefaults.cardColors(
                containerColor = when (monitoringState) {
                    BreathMonitorState.SINGING -> if (isVoiceDetected) Color(0xFF4CAF50) else Color(0xFFFF9800)
                    BreathMonitorState.COMPLETE -> MaterialTheme.colorScheme.primaryContainer
                    else -> MaterialTheme.colorScheme.surfaceVariant
                }
            )
        ) {
            Column(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(24.dp),
                horizontalAlignment = Alignment.CenterHorizontally
            ) {
                Text(
                    text = formatTime(elapsedSeconds),
                    style = MaterialTheme.typography.displayLarge,
                    color = when (monitoringState) {
                        BreathMonitorState.SINGING -> Color.White
                        else -> MaterialTheme.colorScheme.onSurface
                    }
                )

                if (monitoringState == BreathMonitorState.SINGING) {
                    Spacer(modifier = Modifier.height(8.dp))
                    Row(
                        verticalAlignment = Alignment.CenterVertically,
                        horizontalArrangement = Arrangement.spacedBy(8.dp)
                    ) {
                        Box(
                            modifier = Modifier
                                .size(12.dp)
                                .background(
                                    if (isVoiceDetected) Color.White else Color.White.copy(alpha = 0.5f),
                                    CircleShape
                                )
                        )
                        Text(
                            text = if (isVoiceDetected) "Voice detected" else "Silence...",
                            color = Color.White.copy(alpha = 0.8f)
                        )
                    }
                }
            }
        }

        // Level meter
        if (monitoringState != BreathMonitorState.IDLE) {
            Row(
                modifier = Modifier.fillMaxWidth(),
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(8.dp)
            ) {
                Text("Level:", style = MaterialTheme.typography.bodySmall)
                LinearProgressIndicator(
                    progress = { recordingLevel.coerceIn(0f, 1f) },
                    modifier = Modifier
                        .weight(1f)
                        .height(8.dp),
                    color = if (recordingLevel > RMS_THRESHOLD) Color(0xFF4CAF50) else MaterialTheme.colorScheme.primary
                )
            }
        }

        // Best score display
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.SpaceBetween,
            verticalAlignment = Alignment.CenterVertically
        ) {
            Row(verticalAlignment = Alignment.CenterVertically) {
                Text("Best: ", style = MaterialTheme.typography.bodyMedium)
                Text(
                    text = formatTime(bestScore),
                    style = MaterialTheme.typography.titleMedium,
                    color = MaterialTheme.colorScheme.primary
                )
            }
            if (bestScore > 0) {
                TextButton(onClick = { resetBestScore() }) {
                    Text("Reset")
                }
            }
        }

        // Control buttons
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            when (monitoringState) {
                BreathMonitorState.IDLE -> {
                    Button(
                        onClick = { startMonitoring() },
                        modifier = Modifier.weight(1f)
                    ) {
                        Text("Start")
                    }
                }
                BreathMonitorState.COMPLETE -> {
                    Button(
                        onClick = { startMonitoring() },
                        modifier = Modifier.weight(1f)
                    ) {
                        Text("Try Again")
                    }
                }
                else -> {
                    Button(
                        onClick = { stopMonitoring() },
                        modifier = Modifier.weight(1f),
                        colors = ButtonDefaults.buttonColors(
                            containerColor = MaterialTheme.colorScheme.error
                        )
                    ) {
                        Text("Stop")
                    }
                }
            }
        }
    }
}

enum class BreathMonitorState {
    IDLE,
    WAITING_FOR_VOICE,
    SINGING,
    COMPLETE
}

private fun formatTime(seconds: Float): String {
    val mins = (seconds / 60).toInt()
    val secs = seconds % 60
    return if (mins > 0) {
        "%d:%05.2f".format(mins, secs)
    } else {
        "%.2f".format(secs)
    }
}

private fun loadBestScore(context: Context): Float {
    val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    return prefs.getFloat(KEY_BEST_SCORE, 0f)
}

private fun saveBestScore(context: Context, score: Float) {
    val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    prefs.edit().putFloat(KEY_BEST_SCORE, score).apply()
}

File: ios/CalibraSample/CalibraSample/Sections/BreathMonitorSection.swift
import SwiftUI
import calibra
import sonix

private let SILENCE_GRACE_MS: Int64 = 500
private let PREFS_KEY_BEST_SCORE = "breath_monitor_best_score"

enum BreathMonitorState {
    case idle
    case waitingForVoice
    case singing
    case complete
}

/// Breath Monitor using Calibra public API.
///
/// Demonstrates:
/// - `CalibraVAD` (Silero) for voice activity detection
/// - `SonixRecorder` for audio capture
struct BreathMonitorSection: View {
    @State private var monitoringState: BreathMonitorState = .idle
    @State private var elapsedSeconds: Float = 0.0
    @State private var bestScore: Float = UserDefaults.standard.float(forKey: PREFS_KEY_BEST_SCORE)
    @State private var isVoiceDetected = false
    @State private var recordingLevel: Float = 0.0
    @State private var status = "Hold a note as long as you can!"

    @State private var recorder: SonixRecorder?

    // Calibra public API - using VAD for voice detection
    @State private var vad: CalibraVAD?

    @State private var startTimeMs: Int64 = 0
    @State private var lastVoiceTimeMs: Int64 = 0

    var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            Text("Breath Monitor")
                .font(.headline)

            Text(status)
                .font(.subheadline)
                .foregroundColor(.secondary)

            // Main timer display
            timerCard

            // Level meter (volume from Sonix)
            if monitoringState != .idle {
                HStack {
                    Text("Level:")
                        .font(.caption)
                    ProgressView(value: Double(min(max(recordingLevel, 0), 1)))
                        .tint(isVoiceDetected ? .green : .blue)
                }
            }

            // Best score display
            HStack {
                HStack {
                    Text("Best:")
                        .font(.subheadline)
                    Text(formatTime(bestScore))
                        .font(.headline)
                        .foregroundColor(.blue)
                }
                Spacer()
                if bestScore > 0 {
                    Button("Reset") {
                        resetBestScore()
                    }
                    .font(.subheadline)
                }
            }

            // Control buttons
            controlButtons
        }
        .onDisappear {
            stopMonitoring()
            cleanupAudio()
        }
    }

    private var timerCard: some View {
        let backgroundColor: Color = {
            switch monitoringState {
            case .singing:
                return isVoiceDetected ? Color.green : Color.orange
            case .complete:
                return Color.blue.opacity(0.2)
            default:
                return Color(.secondarySystemBackground)
            }
        }()

        let textColor: Color = monitoringState == .singing ? .white : .primary

        return VStack(spacing: 8) {
            Text(formatTime(elapsedSeconds))
                .font(.system(size: 48, weight: .bold, design: .monospaced))
                .foregroundColor(textColor)

            if monitoringState == .singing {
                HStack(spacing: 8) {
                    Circle()
                        .fill(isVoiceDetected ? Color.white : Color.white.opacity(0.5))
                        .frame(width: 12, height: 12)
                    Text(isVoiceDetected ? "Voice detected" : "Silence...")
                        .foregroundColor(.white.opacity(0.8))
                }
            }
        }
        .frame(maxWidth: .infinity)
        .padding(24)
        .background(backgroundColor)
        .cornerRadius(12)
    }

    private var controlButtons: some View {
        HStack {
            switch monitoringState {
            case .idle:
                Button("Start") {
                    startMonitoring()
                }
                .buttonStyle(.borderedProminent)
                .frame(maxWidth: .infinity)

            case .complete:
                Button("Try Again") {
                    startMonitoring()
                }
                .buttonStyle(.borderedProminent)
                .frame(maxWidth: .infinity)

            default:
                Button("Stop") {
                    stopMonitoring()
                }
                .buttonStyle(.borderedProminent)
                .tint(.red)
                .frame(maxWidth: .infinity)
            }
        }
    }

    private func setupAudioIfNeeded() {
        guard vad == nil else { return }

        // Create VAD using Calibra public API with SINGING preset
        vad = CalibraVAD.create(preset: .singing)

        // Create recorder using Sonix
        let tempPath = FileManager.default.temporaryDirectory
            .appendingPathComponent("breath_temp.m4a").path
        recorder = SonixRecorder.create(outputPath: tempPath, format: "m4a", quality: "voice")
    }

    private func cleanupAudio() {
        recorder?.stop()
        recorder?.release()
        recorder = nil

        vad?.release()
        vad = nil
    }

    private func startMonitoring() {
        setupAudioIfNeeded()
        guard let recorder = recorder, let vad = vad else { return }

        // Reset state
        elapsedSeconds = 0
        isVoiceDetected = false
        recordingLevel = 0
        monitoringState = .waitingForVoice
        status = "Start singing when ready..."
        vad.reset()

        // Observe level from Sonix for the meter
        _ = recorder.observeLevel { level in
            recordingLevel = level
        }

        // Collect audio buffers from Sonix for VAD
        Task {
            let hwRate = Int(Sonix.hardwareSampleRate)

            for await buffer in recorder.audioBuffers {
                // Resample to 16kHz for Calibra (expects 16kHz input)
                let samples16k = SonixResampler.resample(
                    samples: buffer.floatSamples,
                    fromRate: hwRate,
                    toRate: 16000
                )

                // Get VAD ratio using Calibra public API
                let ratio = vad.getVADRatio(samples: samples16k)

                // Skip if no complete chunk processed
                if ratio < 0 {
                    continue
                }

                let currentTimeMs = Int64(Date().timeIntervalSince1970 * 1000)
                let hasVoice = ratio > 0.5

                await MainActor.run {
                    switch monitoringState {
                    case .waitingForVoice:
                        if hasVoice {
                            monitoringState = .singing
                            startTimeMs = currentTimeMs
                            lastVoiceTimeMs = currentTimeMs
                            isVoiceDetected = true
                            status = "Keep going!"
                        }

                    case .singing:
                        if hasVoice {
                            lastVoiceTimeMs = currentTimeMs
                            isVoiceDetected = true
                        } else {
                            isVoiceDetected = false
                            let silenceDuration = currentTimeMs - lastVoiceTimeMs

                            if silenceDuration > SILENCE_GRACE_MS {
                                monitoringState = .complete
                                elapsedSeconds = Float(lastVoiceTimeMs - startTimeMs) / 1000.0

                                if elapsedSeconds > bestScore {
                                    bestScore = elapsedSeconds
                                    UserDefaults.standard.set(bestScore, forKey: PREFS_KEY_BEST_SCORE)
                                    status = "New record! \(formatTime(elapsedSeconds))"
                                } else {
                                    status = "Good try! \(formatTime(elapsedSeconds))"
                                }

                                recorder.stop()
                            }
                        }

                        // Update elapsed time
                        if monitoringState == .singing {
                            elapsedSeconds = Float(currentTimeMs - startTimeMs) / 1000.0
                        }

                    default:
                        break
                    }
                }
            }
        }

        recorder.start()
    }

    private func stopMonitoring() {
        recorder?.stop()
        vad?.reset()
        monitoringState = .idle
        status = "Hold a note as long as you can!"
    }

    private func resetBestScore() {
        bestScore = 0
        UserDefaults.standard.set(0, forKey: PREFS_KEY_BEST_SCORE)
    }

    private func formatTime(_ seconds: Float) -> String {
        let mins = Int(seconds / 60)
        let secs = seconds.truncatingRemainder(dividingBy: 60)
        if mins > 0 {
            return String(format: "%d:%05.2f", mins, secs)
        } else {
            return String(format: "%.2f", secs)
        }
    }
}
